File: scripts/generate_grammar_plots.py
# /// script
# requires-python = ">=3.11"
# dependencies = [
#   "altair==5.5.0",
#   "pandas>=2.2,<3",
#   "vl-convert-python>=1.7,<2",
#   "seaborn>=0.13,<1",
#   "matplotlib>=3.9,<4",
# ]
# ///
"""Generate committed static plots for the grammar-of-graphics chapter.

Run from any directory with:

    uv run scripts/generate_grammar_plots.py

The Quarto book displays these SVG files directly. Its Python examples are
therefore visible without starting Jupyter during a normal book build.
"""

from base64 import b64encode
from pathlib import Path
import re

import altair as alt
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
from matplotlib import font_manager
from matplotlib.cm import ScalarMappable
from matplotlib.colors import Normalize
from matplotlib.ticker import EngFormatter

ROOT = Path(__file__).resolve().parents[1]
DATA = ROOT / "data" / "gapminder.csv"
ALTAIR_DIR = ROOT / "media" / "plots" / "grammar" / "altair"
SEABORN_DIR = ROOT / "media" / "plots" / "grammar" / "seaborn"
FONT_REGULAR = ROOT / "media" / "fonts" / "IBMPlexSans-Regular.ttf"
FONT_SEMIBOLD = ROOT / "media" / "fonts" / "IBMPlexSans-SemiBold.ttf"
FONT_WEB = ROOT / "media" / "fonts" / "IBMPlexSans-Latin.woff2"

for font_path in (FONT_REGULAR, FONT_SEMIBOLD):
    font_manager.fontManager.addfont(font_path)

STYLE_FONT = "IBM Plex Sans"
STYLE_GREEN = "#008f6b"
STYLE_GRID = "#EBEBEB"
STYLE_TEXT = "#4D4D4D"
STYLE_X_DOMAIN = (1949, 2011)
STYLE_X_TICKS = list(range(1950, 2011, 5))
STYLE_Y_DOMAIN = (41_000_000, 62_000_000)
STYLE_Y_TICKS = [45_000_000, 50_000_000, 55_000_000, 60_000_000]


def embed_webfont(path: Path) -> None:
    """Embed IBM Plex Sans in an SVG that will be loaded through an img tag."""
    font_data = b64encode(FONT_WEB.read_bytes()).decode("ascii")
    font_face = (
        "<defs><style>"
        "@font-face{font-family:'IBM Plex Sans';font-style:normal;"
        "font-weight:100 700;src:url(data:font/woff2;base64,"
        f"{font_data}) format('woff2');}}"
        "</style></defs>"
    )
    svg = path.read_text(encoding="utf-8")
    svg = re.sub(r"(<svg\b[^>]*>)", rf"\1{font_face}", svg, count=1)
    path.write_text(svg, encoding="utf-8")


def save(
    chart: alt.TopLevelMixin, name: str, *, embed_font: bool = False
) -> None:
    """Save an Altair/Vega-Lite chart as a self-contained SVG."""
    path = ALTAIR_DIR / f"{name}.svg"
    path.parent.mkdir(parents=True, exist_ok=True)
    chart.save(path)
    if embed_font:
        embed_webfont(path)
    print(path.relative_to(ROOT))


def altair_plots(gap: pd.DataFrame) -> None:
    g2007 = gap.query("year == 2007")
    france = gap.query('country == "France"')

    scatter = (
        alt.Chart(g2007)
        .encode(
            x=alt.X("gdp_per_capita:Q", scale=alt.Scale(type="log")),
            y="life_expectancy:Q",
            color="continent:N",
            size=alt.Size("population:Q", legend=None),
            tooltip=["country:N", "population:Q"],
        )
        .mark_circle()
        .properties(width=500, height=300)
    )
    save(scatter, "scatter-2007")

    continents = (
        alt.Chart(g2007)
        .encode(
            y=alt.Y("continent:N", title=None),
            color=alt.Color("continent:N", legend=None),
        )
        .mark_square(size=100)
    )
    save(continents, "continent-squares")

    france_line = (
        alt.Chart(france)
        .encode(
            x=alt.X("year:Q", title="year", axis=alt.Axis(format="d")),
            y=alt.Y(
                "population:Q",
                scale=alt.Scale(zero=False),
                axis=alt.Axis(format="~s"),
            ),
        )
        .mark_line()
        .properties(width=500)
    )
    save(france_line, "france-population")

    median_gdp = (
        alt.Chart(g2007)
        .encode(
            x=alt.X(
                "median(gdp_per_capita):Q",
                title="median GDP per capita",
                axis=alt.Axis(format="~s"),
            ),
            y=alt.Y("continent:N", sort="-x"),
            color="continent:N",
        )
        .mark_bar()
        .properties(width=500)
    )
    save(median_gdp, "median-gdp-continent")

    country_count = (
        alt.Chart(gap)
        .encode(
            x=alt.X("distinct(country):Q", title="number of countries"),
            y=alt.Y("continent:N", sort="-x"),
            color=alt.Color("continent:N", legend=None),
        )
        .mark_bar()
    )
    save(country_count, "country-count-continent")

    dispersion = (
        alt.Chart(g2007)
        .encode(
            x=alt.X(
                "mean(gdp_per_capita):Q",
                axis=alt.Axis(format="~s"),
                title="mean GDP per capita",
            ),
            y=alt.Y("continent:N", sort="-x"),
            color=alt.Color(
                "stdev(gdp_per_capita):Q",
                title="standard deviation",
                scale=alt.Scale(domain=(0, 20_000), scheme="blues"),
            ),
        )
        .mark_bar()
    )
    save(dispersion, "gdp-dispersion-continent")

    boxplot = (
        alt.Chart(g2007)
        .encode(
            x=alt.X(
                "gdp_per_capita:Q",
                axis=alt.Axis(format="~s"),
                title="GDP per capita",
            ),
            y="continent:N",
            tooltip="country:N",
        )
        .mark_boxplot()
    )
    save(boxplot, "gdp-boxplot")

    histogram_base = (
        alt.Chart(g2007)
        .encode(
            x=alt.X(
                "gdp_per_capita:Q",
                bin=alt.Bin(maxbins=30),
                title="GDP per capita",
            ),
            y=alt.Y("count():Q", title="number of countries"),
        )
        .mark_bar()
        .properties(width=300, height=200)
    )
    save(histogram_base | histogram_base.encode(color="continent:N"), "gdp-histograms")

    population_base = (
        alt.Chart(g2007)
        .encode(
            x=alt.X(
                "sum(population):Q",
                title="total population in 2007",
                axis=alt.Axis(format="~s"),
            ),
            color=alt.Color("continent:N", legend=None),
        )
        .mark_bar(size=10)
        .properties(width=280)
    )
    population_layout = (
        population_base.encode(y=alt.Y("continent:N", title=None))
        | population_base.encode(
            x=alt.X("population:Q", axis=alt.Axis(format="~s")),
            y=alt.Y("continent:N", title=None),
        ).mark_point()
    ) & population_base.properties(width=680)
    save(population_layout, "population-layout")

    large_europe = (
        alt.Chart(gap)
        .encode(
            x=alt.X("year:T", title="year"),
            y=alt.Y("population:Q", axis=alt.Axis(format="~s")),
            color=alt.Color("country:N", title="country"),
            order=alt.Order("mean_pop:Q", sort="descending"),
        )
        .mark_area()
        .transform_joinaggregate(mean_pop="mean(population)", groupby=["country"])
        .transform_filter("datum.continent == 'Europe' && datum.mean_pop > 50000000")
        .properties(width=480, height=240)
    )
    save(large_europe, "large-europe-population")

    ranking_base = (
        alt.Chart(g2007)
        .mark_bar(size=10)
        .encode(
            y=alt.Y("country:N", sort="-x", title="country"),
            color=alt.Color("continent:N", legend=None),
        )
        .transform_window(
            rank_pop="rank(population)",
            sort=[alt.SortField("population", order="descending")],
        )
        .transform_window(
            rank_gdp="rank(gdp_per_capita)",
            sort=[alt.SortField("gdp_per_capita", order="descending")],
        )
        .properties(width=300, height=220)
    )
    ranking = ranking_base.encode(
        x=alt.X("population:Q", axis=alt.Axis(format="~s"))
    ).transform_filter("datum.rank_pop <= 10") | ranking_base.encode(
        x=alt.X(
            "gdp_per_capita:Q",
            axis=alt.Axis(format="~s"),
            title="GDP per capita",
        )
    ).transform_filter("datum.rank_gdp <= 10")
    save(ranking, "top10-population-gdp")

    year_slider = alt.binding_range(min=1952, max=2007, step=5, name="year:")
    year_selector = alt.selection_point(
        name="year_selection",
        fields=["year"],
        bind=year_slider,
        value=[{"year": 2007}],
    )
    bubble = (
        alt.Chart(gap)
        .encode(
            x=alt.X(
                "gdp_per_capita:Q",
                scale=alt.Scale(type="log", domain=(300, 1e5)),
                title="GDP per capita",
            ),
            y=alt.Y(
                "life_expectancy:Q",
                scale=alt.Scale(domain=(20, 90)),
                title="life expectancy",
            ),
            color="continent:N",
            size=alt.Size(
                "population:Q",
                scale=alt.Scale(domain=(2e5, 1.4e9), type="log"),
                legend=None,
            ),
            tooltip="country:N",
        )
        .transform_filter(year_selector)
        .properties(width=420, height=320)
        .mark_circle()
        .add_params(year_selector)
    )
    save(bubble, "rosling-bubbles-2007")

    styled = (
        alt.Chart(france)
        .encode(
            x=alt.X(
                "year:Q",
                scale=alt.Scale(domain=STYLE_X_DOMAIN, nice=False),
                axis=alt.Axis(
                    title=None,
                    values=STYLE_X_TICKS,
                    format="d",
                    grid=True,
                    gridColor=STYLE_GRID,
                    gridWidth=1.43,
                    domain=False,
                    ticks=False,
                    labelColor=STYLE_TEXT,
                    labelFont=STYLE_FONT,
                    labelFontSize=12,
                    labelFlush=True,
                    labelOverlap=False,
                    labelPadding=7,
                ),
            ),
            y=alt.Y(
                "population:Q",
                scale=alt.Scale(domain=STYLE_Y_DOMAIN, nice=False),
                axis=alt.Axis(
                    title=None,
                    values=STYLE_Y_TICKS,
                    format="~s",
                    grid=True,
                    gridColor=STYLE_GRID,
                    gridWidth=1.43,
                    domain=False,
                    ticks=False,
                    labelColor=STYLE_TEXT,
                    labelFont=STYLE_FONT,
                    labelFontSize=12,
                    labelPadding=0,
                ),
            ),
        )
        .mark_line(color=STYLE_GREEN, strokeWidth=2.84)
        .properties(
            title="Population of France",
            width=576,
            height=346,
            autosize=alt.AutoSizeParams(type="fit", contains="padding"),
        )
        .configure_title(
            anchor="start",
            color="#000000",
            font=STYLE_FONT,
            fontSize=18,
            fontWeight=600,
            dx=33.4,
            offset=8,
        )
        .configure_view(stroke=None)
    )
    save(styled, "france-population-styled", embed_font=True)


def save_mpl(fig: plt.Figure, name: str, *, tight: bool = True) -> None:
    """Save a matplotlib/seaborn figure as SVG and release its resources."""
    path = SEABORN_DIR / f"{name}.svg"
    path.parent.mkdir(parents=True, exist_ok=True)
    fig.savefig(path, format="svg", bbox_inches="tight" if tight else None)
    plt.close(fig)
    print(path.relative_to(ROOT))


def save_mpl_png(fig: plt.Figure, name: str) -> None:
    """Save a browser-portable, double-density matplotlib PNG."""
    path = SEABORN_DIR / f"{name}.png"
    path.parent.mkdir(parents=True, exist_ok=True)
    fig.savefig(path, format="png", dpi=192, facecolor="white")
    plt.close(fig)
    print(path.relative_to(ROOT))


def seaborn_plots(gap: pd.DataFrame) -> None:
    sns.set_theme(style="whitegrid")
    g2007 = gap.query("year == 2007")
    france = gap.query('country == "France"')
    continents = sorted(g2007["continent"].unique())
    palette = dict(zip(continents, sns.color_palette("tab10", len(continents))))
    formatter = EngFormatter(sep="")

    fig, ax = plt.subplots(figsize=(6.4, 4.2))
    sns.scatterplot(
        data=g2007,
        x="gdp_per_capita",
        y="life_expectancy",
        hue="continent",
        size="population",
        sizes=(20, 500),
        palette=palette,
        ax=ax,
    )
    ax.set(xscale="log", xlabel="GDP per capita", ylabel="life expectancy")
    sns.move_legend(ax, "center left", bbox_to_anchor=(1, 0.5), frameon=False)
    save_mpl(fig, "scatter-2007")

    fig, ax = plt.subplots(figsize=(5.4, 2.8))
    sns.scatterplot(
        data=g2007.assign(x=0),
        x="x",
        y="continent",
        hue="continent",
        marker="s",
        s=110,
        palette=palette,
        legend=False,
        ax=ax,
    )
    ax.set(xlabel=None, ylabel=None, xticks=[])
    sns.despine(ax=ax, bottom=True)
    save_mpl(fig, "continent-squares")

    fig, ax = plt.subplots(figsize=(6.2, 3.6))
    sns.lineplot(data=france, x="year", y="population", ax=ax)
    ax.yaxis.set_major_formatter(formatter)
    ax.set(xlabel="year", ylabel="population")
    save_mpl(fig, "france-population")

    median_order = (
        g2007.groupby("continent")["gdp_per_capita"]
        .median()
        .sort_values()
        .index
    )
    fig, ax = plt.subplots(figsize=(6.2, 3.8))
    sns.barplot(
        data=g2007,
        x="gdp_per_capita",
        y="continent",
        hue="continent",
        estimator="median",
        errorbar=None,
        order=median_order,
        palette=palette,
        legend=False,
        ax=ax,
    )
    ax.xaxis.set_major_formatter(formatter)
    ax.set(xlabel="median GDP per capita", ylabel=None)
    save_mpl(fig, "median-gdp-continent")

    country_count = (
        gap.groupby("continent", as_index=False)
        .agg(countries=("country", "nunique"))
        .sort_values("countries")
    )
    fig, ax = plt.subplots(figsize=(6.2, 3.8))
    sns.barplot(
        data=country_count,
        x="countries",
        y="continent",
        hue="continent",
        palette=palette,
        legend=False,
        ax=ax,
    )
    ax.set(xlabel="number of countries", ylabel=None)
    save_mpl(fig, "country-count-continent")

    dispersion = (
        g2007.groupby("continent", as_index=False)["gdp_per_capita"]
        .agg(mean="mean", stdev="std")
        .sort_values("mean")
    )
    norm = Normalize(0, 20_000)
    cmap = plt.get_cmap("Blues")
    dispersion_palette = {
        row.continent: cmap(norm(row.stdev)) for row in dispersion.itertuples()
    }
    fig, ax = plt.subplots(figsize=(6.8, 4.0))
    sns.barplot(
        data=dispersion,
        x="mean",
        y="continent",
        hue="continent",
        palette=dispersion_palette,
        legend=False,
        ax=ax,
    )
    fig.colorbar(
        ScalarMappable(norm=norm, cmap=cmap),
        ax=ax,
        label="standard deviation",
    )
    ax.xaxis.set_major_formatter(formatter)
    ax.set(xlabel="mean GDP per capita", ylabel=None)
    save_mpl(fig, "gdp-dispersion-continent")

    fig, ax = plt.subplots(figsize=(6.4, 4.0))
    sns.boxplot(data=g2007, x="gdp_per_capita", y="continent", ax=ax)
    ax.xaxis.set_major_formatter(formatter)
    ax.set(xlabel="GDP per capita", ylabel=None)
    save_mpl(fig, "gdp-boxplot")

    fig, axes = plt.subplots(1, 2, figsize=(10.5, 3.8), sharey=True)
    sns.histplot(data=g2007, x="gdp_per_capita", bins=30, ax=axes[0])
    sns.histplot(
        data=g2007,
        x="gdp_per_capita",
        hue="continent",
        palette=palette,
        multiple="stack",
        bins=30,
        ax=axes[1],
    )
    for ax in axes:
        ax.set(xlabel="GDP per capita", ylabel="number of countries")
    save_mpl(fig, "gdp-histograms")

    totals = (
        g2007.groupby("continent", as_index=False)["population"]
        .sum()
        .sort_values("population")
    )
    fig = plt.figure(figsize=(11.0, 6.0), layout="constrained")
    grid = fig.add_gridspec(2, 2)
    ax_total = fig.add_subplot(grid[0, 0])
    ax_points = fig.add_subplot(grid[0, 1])
    ax_stack = fig.add_subplot(grid[1, :])
    sns.barplot(
        data=totals,
        x="population",
        y="continent",
        hue="continent",
        palette=palette,
        legend=False,
        ax=ax_total,
    )
    sns.scatterplot(
        data=g2007,
        x="population",
        y="continent",
        hue="continent",
        palette=palette,
        legend=False,
        ax=ax_points,
    )
    left = 0
    for row in totals.itertuples():
        ax_stack.barh(
            ["total"],
            row.population,
            left=left,
            color=palette[row.continent],
            label=row.continent,
        )
        left += row.population
    ax_stack.legend(ncol=3, frameon=False)
    for ax in (ax_total, ax_points, ax_stack):
        ax.xaxis.set_major_formatter(formatter)
        ax.set_xlabel("population")
    save_mpl(fig, "population-layout")

    mean_population = gap.groupby("country")["population"].transform("mean")
    large_europe = gap[(gap["continent"] == "Europe") & (mean_population > 50e6)]
    wide = large_europe.pivot(index="year", columns="country", values="population")
    fig, ax = plt.subplots(figsize=(7.6, 4.2))
    colours = sns.color_palette("tab10", wide.shape[1])
    ax.stackplot(
        wide.index,
        *[wide[column] for column in wide.columns],
        labels=wide.columns,
        colors=colours,
    )
    ax.yaxis.set_major_formatter(formatter)
    ax.set(xlabel="year", ylabel="population")
    ax.legend(loc="center left", bbox_to_anchor=(1, 0.5), frameon=False)
    save_mpl(fig, "large-europe-population")

    top_population = g2007.nlargest(10, "population").sort_values("population")
    top_gdp = g2007.nlargest(10, "gdp_per_capita").sort_values("gdp_per_capita")
    fig, axes = plt.subplots(1, 2, figsize=(11.0, 4.8))
    sns.barplot(
        data=top_population,
        x="population",
        y="country",
        hue="continent",
        palette=palette,
        legend=False,
        ax=axes[0],
    )
    sns.barplot(
        data=top_gdp,
        x="gdp_per_capita",
        y="country",
        hue="continent",
        palette=palette,
        legend=False,
        ax=axes[1],
    )
    axes[0].set(xlabel="population", ylabel=None)
    axes[1].set(xlabel="GDP per capita", ylabel=None)
    for ax in axes:
        ax.xaxis.set_major_formatter(formatter)
    save_mpl(fig, "top10-population-gdp")

    with plt.rc_context({
        "font.family": STYLE_FONT,
        "font.size": 8.8,
        "axes.titlesize": 13.2,
        "axes.titleweight": 600,
        "axes.titlecolor": "#000000",
        "xtick.color": STYLE_TEXT,
        "ytick.color": STYLE_TEXT,
        "svg.fonttype": "none",
    }):
        fig, ax = plt.subplots(figsize=(6.0, 3.6))
        fig.subplots_adjust(
            left=29.25 / 432,
            right=420 / 432,
            bottom=(259.2 - 242.25) / 259.2,
            top=(259.2 - 23.25) / 259.2,
        )
        sns.lineplot(
            data=france,
            x="year",
            y="population",
            color=STYLE_GREEN,
            linewidth=2.13,
            solid_capstyle="butt",
            ax=ax,
        )
        ax.set_title("Population of France", loc="left", pad=8.5)
        ax.set(
            xlabel=None,
            ylabel=None,
            xlim=STYLE_X_DOMAIN,
            ylim=STYLE_Y_DOMAIN,
            xticks=STYLE_X_TICKS,
            yticks=STYLE_Y_TICKS,
        )
        ax.yaxis.set_major_formatter(formatter)
        ax.grid(True, which="major", color=STYLE_GRID, linewidth=1.07)
        for gridline in (*ax.get_xgridlines(), *ax.get_ygridlines()):
            gridline.set_solid_capstyle("butt")
        ax.tick_params(axis="both", length=0, labelsize=8.8, pad=5)
        sns.despine(ax=ax, left=True, bottom=True)
        save_mpl_png(fig, "france-population-styled")


def main() -> None:
    gap = pd.read_csv(DATA)
    altair_plots(gap)
    seaborn_plots(gap)


if __name__ == "__main__":
    main()

File: scripts/generate_grammar_plots.R
#!/usr/bin/env Rscript
# Generate committed static ggplot2 plots for the grammar-of-graphics chapter.
#
# Run with the rig-managed R (see HOW-TO.md for the one-time setup):
#
#   Rscript scripts/generate_grammar_plots_ggplot2.R
#
# If `Rscript` on the PATH resolves to an old R, invoke the new one directly:
#
#   /Library/Frameworks/R.framework/Versions/4.6/Resources/bin/Rscript \
#     scripts/generate_grammar_plots_ggplot2.R

library(ggplot2)
library(dplyr)
library(tidyr)
library(scales)
library(svglite)
library(patchwork)
library(readr)

root <- normalizePath(file.path(getwd(), ".."), winslash = "/")
if (!file.exists(file.path(root, "data", "gapminder.csv"))) {
  root <- getwd()
}
data_path <- file.path(root, "data", "gapminder.csv")
out_dir <- file.path(root, "media", "plots", "grammar", "ggplot2")
dir.create(out_dir, showWarnings = FALSE, recursive = TRUE)

gap <- read_csv(data_path, show_col_types = FALSE)
g2007 <- gap |> filter(year == 2007)
france <- gap |> filter(country == "France")

font_regular <- file.path(root, "media", "fonts", "IBMPlexSans-Regular.ttf")
font_semibold <- file.path(root, "media", "fonts", "IBMPlexSans-SemiBold.ttf")
systemfonts::register_font(
  "IBM Plex Sans",
  plain = font_regular,
  bold = font_semibold
)

continents <- sort(unique(g2007$continent))
palette <- setNames(
  scales::hue_pal()(length(continents)), continents
)

# SI-style formatting (60400000 -> 60M), like d3's ~s.
si_format <- function() scales::label_number(scale_cut = scales::cut_short_scale())
theme_page <- function() {
  theme_minimal(base_size = 11) +
    theme(
      plot.title = element_text(hjust = 0, face = "bold"),
      panel.grid.minor = element_blank()
    )
}

save_plot <- function(
  plot, name, width = 6.4, height = 4.2, apply_page_theme = TRUE
) {
  if (apply_page_theme) plot <- plot + theme_page()
  path <- file.path(out_dir, paste0(name, ".svg"))
  ggsave(
    path, plot, device = svglite::svglite,
    width = width, height = height, units = "in"
  )
  cat(relpath(path), "\n", sep = "")
}

save_plot_png <- function(plot, name, width = 6.0, height = 3.6) {
  path <- file.path(out_dir, paste0(name, ".png"))
  ggsave(
    path, plot, device = ragg::agg_png,
    width = width, height = height, units = "in", dpi = 192,
    background = "white"
  )
  cat(relpath(path), "\n", sep = "")
}

relpath <- function(path) {
  gsub("^./", "", path)
}

# 1. Rosling scatter, 2007 -----------------------------------------------
p_scatter <- ggplot(g2007, aes(
  x = gdp_per_capita, y = life_expectancy,
  colour = continent, size = population
)) +
  geom_point(alpha = 0.85) +
  scale_x_log10(labels = si_format()) +
  scale_size_continuous(range = c(1.5, 14), guide = "none") +
  labs(x = "GDP per capita", y = "life expectancy", colour = "continent")
save_plot(p_scatter, "scatter-2007")

# 2. Nominal continent squares ------------------------------------------
p_squares <- ggplot(g2007 |> mutate(x = 0), aes(x, continent, colour = continent)) +
  geom_point(shape = 15, size = 6) +
  scale_colour_manual(values = palette, guide = "none") +
  scale_x_continuous(breaks = NULL) +
  labs(x = NULL, y = NULL)
save_plot(p_squares, "continent-squares", width = 5.4, height = 2.8)

# 4. France population line ---------------------------------------------
p_france <- ggplot(france, aes(year, population)) +
  geom_line() +
  scale_y_continuous(labels = si_format()) +
  labs(x = "year", y = "population")
save_plot(p_france, "france-population", width = 6.2, height = 3.6)

# 5. Median GDP per continent -------------------------------------------
p_median <- ggplot(g2007, aes(
  gdp_per_capita,
  reorder(continent, gdp_per_capita, FUN = median),
  fill = continent
)) +
  stat_summary(fun = median, geom = "bar", orientation = "y") +
  scale_fill_manual(values = palette, guide = "none") +
  scale_x_continuous(labels = si_format()) +
  labs(x = "median GDP per capita", y = NULL)
save_plot(p_median, "median-gdp-continent", width = 6.2, height = 3.8)

# 6. Distinct country count per continent -------------------------------
country_count <- gap |>
  group_by(continent) |>
  summarise(countries = n_distinct(country), .groups = "drop") |>
  mutate(continent = reorder(continent, countries))
p_count <- ggplot(country_count, aes(countries, continent, fill = continent)) +
  geom_col() +
  scale_fill_manual(values = palette, guide = "none") +
  labs(x = "number of countries", y = NULL)
save_plot(p_count, "country-count-continent", width = 6.2, height = 3.8)

# 7. Mean GDP coloured by standard deviation ----------------------------
dispersion <- g2007 |>
  group_by(continent) |>
  summarise(mean = mean(gdp_per_capita), stdev = sd(gdp_per_capita), .groups = "drop") |>
  mutate(continent = reorder(continent, mean))
p_disp <- ggplot(dispersion, aes(mean, continent, fill = stdev)) +
  geom_col() +
  scale_fill_distiller(palette = "Blues", direction = 1, limits = c(0, 20000)) +
  scale_x_continuous(labels = si_format()) +
  labs(x = "mean GDP per capita", y = NULL, fill = "standard deviation") +
  theme(legend.position = "right")
save_plot(p_disp, "gdp-dispersion-continent", width = 6.8, height = 4.0)

# 8. GDP per capita box plots -------------------------------------------
p_box <- ggplot(g2007, aes(gdp_per_capita, continent)) +
  geom_boxplot() +
  scale_x_continuous(labels = si_format()) +
  labs(x = "GDP per capita", y = NULL)
save_plot(p_box, "gdp-boxplot", width = 6.4, height = 4.0)

# 9. Histograms: plain and stacked by continent -------------------------
p_hist1 <- ggplot(g2007, aes(gdp_per_capita)) +
  geom_histogram(bins = 30, fill = "grey50") +
  labs(x = "GDP per capita", y = "number of countries")
p_hist2 <- ggplot(g2007, aes(gdp_per_capita, fill = continent)) +
  geom_histogram(bins = 30, position = "stack") +
  scale_fill_manual(values = palette) +
  labs(x = "GDP per capita", y = "number of countries")
save_plot(p_hist1 | p_hist2, "gdp-histograms", width = 10.5, height = 3.8)

# 10. Population composition (totals, points, stacked total) ------------
totals <- g2007 |>
  group_by(continent) |>
  summarise(population = sum(population), .groups = "drop") |>
  mutate(continent = reorder(continent, population))
p_totals <- ggplot(totals, aes(population, continent, fill = continent)) +
  geom_col() +
  scale_fill_manual(values = palette, guide = "none") +
  scale_x_continuous(labels = si_format()) +
  labs(x = "population", y = NULL)
p_points <- ggplot(g2007, aes(population, continent, colour = continent)) +
  geom_point(alpha = 0.6) +
  scale_colour_manual(values = palette, guide = "none") +
  scale_x_continuous(labels = si_format()) +
  labs(x = "population", y = NULL)
stack_frame <- totals |>
  mutate(category = "total") |>
  select(category, continent, population)
p_stack <- ggplot(stack_frame, aes(population, category, fill = continent)) +
  geom_col() +
  scale_fill_manual(values = palette) +
  scale_x_continuous(labels = si_format()) +
  labs(x = "population", y = NULL)
composition <- (p_totals | p_points) / p_stack + plot_layout(heights = c(1, 1))
save_plot(composition, "population-layout", width = 11.0, height = 6.0)

# 11. Stacked area, large European countries ----------------------------
mean_pop <- gap |>
  group_by(country) |>
  summarise(mean_pop = mean(population), .groups = "drop")
large_europe <- gap |>
  filter(continent == "Europe") |>
  inner_join(mean_pop, by = "country") |>
  filter(mean_pop > 5e7) |>
  mutate(country = reorder(country, -mean_pop))
p_area <- ggplot(large_europe, aes(year, population, fill = country)) +
  geom_area() +
  scale_fill_manual(values = scales::hue_pal()(n_distinct(large_europe$country))) +
  scale_y_continuous(labels = si_format()) +
  labs(x = "year", y = "population", fill = "country") +
  theme(legend.position = "right")
save_plot(p_area, "large-europe-population", width = 7.6, height = 4.2)

# 12. Top 10 by population and by GDP per capita ------------------------
top_pop <- g2007 |>
  slice_max(population, n = 10) |>
  mutate(country = reorder(country, population))
top_gdp <- g2007 |>
  slice_max(gdp_per_capita, n = 10) |>
  mutate(country = reorder(country, gdp_per_capita))
p_top_pop <- ggplot(top_pop, aes(population, country, fill = continent)) +
  geom_col() +
  scale_fill_manual(values = palette, guide = "none") +
  scale_x_continuous(labels = si_format()) +
  labs(x = "population", y = NULL)
p_top_gdp <- ggplot(top_gdp, aes(gdp_per_capita, country, fill = continent)) +
  geom_col() +
  scale_fill_manual(values = palette, guide = "none") +
  scale_x_continuous(labels = si_format()) +
  labs(x = "GDP per capita", y = NULL)
save_plot(p_top_pop | p_top_gdp, "top10-population-gdp", width = 11.0, height = 4.8)

# 13. Static Rosling bubble chart, initial 2007 state -------------------
p_bubble <- ggplot(g2007, aes(
  gdp_per_capita, life_expectancy,
  colour = continent, size = population
)) +
  geom_point(alpha = 0.85) +
  scale_x_log10(limits = c(300, 1e5), labels = si_format()) +
  scale_y_continuous(limits = c(20, 90)) +
  scale_size_continuous(
    range = c(1.5, 18), breaks = c(1e6, 1e7, 1e8, 1e9), guide = "none"
  ) +
  labs(x = "GDP per capita", y = "life expectancy", colour = "continent")
save_plot(p_bubble, "rosling-bubbles-2007", width = 6.4, height = 4.8)

# 14. Styled France population line -------------------------------------
p_styled <- ggplot(france, aes(year, population)) +
  geom_line(colour = "#008f6b", linewidth = 1) +
  scale_x_continuous(
    breaks = seq(1950, 2010, by = 5),
    limits = c(1949, 2011),
    expand = expansion(mult = 0)
  ) +
  scale_y_continuous(
    breaks = c(45e6, 50e6, 55e6, 60e6),
    limits = c(41e6, 62e6),
    labels = si_format(),
    expand = expansion(mult = 0)
  ) +
  labs(title = "Population of France", x = NULL, y = NULL) +
  theme_minimal(base_family = "IBM Plex Sans", base_size = 11) +
  theme(
    plot.title = element_text(
      family = "IBM Plex Sans", face = "bold", size = 13.2,
      hjust = 0, margin = margin(b = 6)
    ),
    axis.text = element_text(
      family = "IBM Plex Sans", colour = "#4D4D4D", size = 8.8
    ),
    axis.ticks = element_blank(),
    panel.grid.major = element_line(colour = "#EBEBEB", linewidth = 0.5),
    panel.grid.minor = element_blank(),
    plot.margin = margin(t = 5.29, r = 12.02, b = 4.03, l = 6.64)
  )
save_plot_png(p_styled, "france-population-styled")
